src/Spice86.Core/Emulator/OperatingSystem/DosPathResolver.cs

```diff
@@ -0,0 +1,235 @@
+        if (parent == null) {
+            return null;
+        }
+
+        return new DirectoryInfo(parent).GetDirectories(directoryInfo.Name, new EnumerationOptions { MatchCasing = MatchCasing.CaseInsensitive }).FirstOrDefault()?.FullName;
+    }
+
+    /// <summary>
+    /// Searches for the path on disk, and returns the first matching file or folder path.
+    /// </summary>
+    /// <param name="caseInsensitivePath">The case insensitive file path</param>
+    /// <param name="convertParentOnly"></param>
+    /// <returns>The absolute host file path</returns>
+    private string? ToCaseSensitiveFileName(string? caseInsensitivePath, bool convertParentOnly) {
+        if (string.IsNullOrWhiteSpace(caseInsensitivePath)) {
+            return null;
+        }
+
+        string fileToProcess = ConvertUtils.ToSlashPath(caseInsensitivePath);
+        string? parentDir = Path.GetDirectoryName(fileToProcess);
+        if (File.Exists(fileToProcess) || 
+            Directory.Exists(fileToProcess) ||
+            convertParentOnly) {
+            // file exists or root reached, no need to go further. Path found.
+            return caseInsensitivePath;
+        }
+
+        string? parent = ToCaseSensitiveFileName(parentDir, convertParentOnly);
+        if (parent == null) {
+            // End of recursion, root reached
+            return null;
+        }
+
+        // Now that parent is for sure on the disk, let's find the current file
+        try {
+            string? fileNameOnFileSystem = TryGetFullHostFileName(caseInsensitivePath);
+            if (!string.IsNullOrWhiteSpace(fileNameOnFileSystem)) {
+                return fileNameOnFileSystem;
+            }
+            Regex fileToProcessRegex = FileSpecToRegex(Path.GetFileName(fileToProcess));
+            if (Directory.Exists(parent)) {
+                return Array.Find(Directory.GetFiles(parent), 
+                    x => fileToProcessRegex.IsMatch(x));
+            }
+        } catch (IOException e) {
+            e.Demystify();
+            if (_loggerService.IsEnabled(Serilog.Events.LogEventLevel.Warning)) {
+                _loggerService.Warning(e, "Error while checking file {CaseInsensitivePath}: {Exception}", caseInsensitivePath, e);
+            }
+        }
+
+        return null;
+    }
+    
+    /// <summary>
+    /// Converts a dosVirtualDevices filespec to a regex pattern
+    /// </summary>
+    /// <param name="fileSpec">The DOS filespec</param>
+    /// <returns>The regex pattern</returns>
+    private static Regex FileSpecToRegex(string fileSpec) {
+        string regex = fileSpec.ToLowerInvariant();
+        regex = regex.Replace(".", "[.]");
+        regex = regex.Replace("?", ".");
+        regex = regex.Replace("*", ".*");
+        return new Regex(regex);
+    }
+    
+    /// <inheritdoc />
+    public string? ToHostCaseSensitiveFullName(string dosFileName, bool convertParentOnly) {
```

Comment:
Could we have two separate methods, rather than a boolean flag?
a `ToHostCaseSensitiveFullName` and `ToHostCaseSensitiveParentName` or something?

Reply:
I'd like to remove that boolean too.

Reply:
Fixed.